oeps/oep-0023-theming.rst

```diff
@@ -0,0 +1,204 @@
+customization, most of the core team (with the notable exception of the
+White Label team) does not. Domain knowledge is sparse and siloed and
+documentation is poorly maintained.
+
+Without an officially supported API, core contributors are largely
+unaware of what they can and cannot change. Fear of breaking existing
+theming systems paralyzes the core team’s ability to make crucial
+evolutionary changes to the platform’s frontend. Because existing
+theming systems rely so heavily on legacy Python template rendering, the
+platform frontend has been unable to evolve towards a modern,
+developer-friendly, client-rendered architecture, frustrating core
+contributors and community members alike.
+
+Ironically, this same lack of awareness means undocumented breaking
+changes to theming source files still occur with every release. Without
+an official API, the Open edX community has no easy way to keep up with
+breaking changes to templates and source files, leading to widespread
+bugs and frustration. Many community members have abandoned Open edX’s
+frontend entirely in favor of building their own, indicating an absolute
+failure of existing theming systems for their use cases.
+
+Decision
+========
+
+To guarantee continued core contributor support and full transparency to
+the open source community, the Open edX platform must officially support
+a single well-defined system for interface customization. Official
+support for interface customization requires a more limited scope than
+has been allowed by prior theming systems. However, isolating theming
+mechanisms to a single system reduces cognitive load for both platform
+contributors and maintainers, facilitates communication of breaking
+changes, and enables bolder architectural innovation.
+
+Goals
+-----
+
+1. The Open edX core team will officially support a theming system that
+      enables contributors to easily customize look-and-feel and
+      branding elements across user-facing components of all Open edX
+      applications.
+
+2. This theming system will surface an API defining which elements can
+      be modified and how. As outlined in OEP-16 [7], this API will be
+      based on the Bootstrap framework.
+
+3. As consumers of the new theming API, the core contributor team will
+      make every reasonable effort to keep breaking changes to an
+      absolute minimum. If breaking changes must occur, the core
+      contributor team will enumerate them with every named release of
+      the Open edX platform.
+
+4. The core contributor team will provide official documentation,
+      examples, and tools to help other core contributors and open
+      source community members develop themes.
+
+5. Further modification to UI elements or content will not be officially
+      supported within the scope of this theming system. A system to
+      enable content modification may eventually be developed and
+      supported, but its definition is outside the scope of this
+      proposal.
```

Comment:
Content modification is a very common requirement, so I agree with Braden it should be clear how that's going to be supported before the frontend starts being rewritten in React. While content modification with current comprehensive theming system is often kind of messy, modifying a React based interface is going to be even worse if it's not clearly thought out.

Reply:
@bradenmacdonald I think this is an excellent approach and would be interested in seeing it grow into an OEP! The deprecation path is a broader conversation -- I'll leave a more general comment below.

Reply:
> I think this is an excellent approach and would be interested in seeing it grow into an OEP! 

@arizzitano Great! Do you know if you or anyone at edX is interested in taking that on? I might be able to find someone to do so but I don't have capacity to develop it further myself right at the moment.

CC @mtyaka in case you're interested :)